core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -378,6 +377,38 @@ private[spark] class SparkSubmit extends Logging {
     val hadoopConf = conf.getOrElse(SparkHadoopUtil.newConfiguration(sparkConf))
     val targetDir = Utils.createTempDir()
 
+    if (clusterManager == EXTERNAL) {
```

Comment:
Since we are past

      // Resolve maven dependencies if there are any and add classpath to jars.

here, packages defined via `--packages` should have been resolved and added to `args.jars` by now, so cluster manager class should be available without the need of `EXTERNAL_CLUSTER_MANAGER_JARS`?

If you move this block past

    // At this point, we have attempted to download all remote resources.

Then non-local jars (like `--jars hdfs://`) will be available in client deploy mode as well.